conngen/conngen.cpp

```diff
@@ -37,185 +34,312 @@
 // Includes from sli:
 #include "token.h"
 
+namespace nest
+{
+
 void
-nest::cg_connect( nest::ConnectionGeneratorDatum& cg,
-  const index source_id,
-  const index target_id,
+cg_connect( ConnectionGeneratorDatum& cg,
+  const GIDCollection& source_gids,
+  const GIDCollection& target_gids,
   const DictionaryDatum& params_map,
   const Name& synmodel_name )
 {
-  Subnet* sources =
-    dynamic_cast< Subnet* >( kernel().node_manager.get_node( source_id ) );
-  if ( sources == NULL )
-  {
-    LOG( M_ERROR, "CGConnect_cg_i_i_D_l", "sources must be a subnet." );
-    throw SubnetExpected();
-  }
-  if ( not sources->is_homogeneous() )
-  {
-    LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "sources must be a homogeneous subnet." );
-    throw BadProperty();
-  }
-  if ( dynamic_cast< Subnet* >( *sources->local_begin() ) )
+  const Token synmodel =
+    kernel().model_manager.get_synapsedict()->lookup( synmodel_name );
+  if ( synmodel.empty() )
   {
-    LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "Only 1-dim subnets are supported as sources." );
-    throw BadProperty();
+    throw UnknownSynapseType( synmodel_name.toString() );
   }
+  const index synmodel_id = static_cast< index >( synmodel );
+
+  RangeSet source_ranges, target_ranges;
+
+  cg_get_ranges( source_ranges, source_gids );
+  cg_get_ranges( target_ranges, target_gids );
+  cg_set_masks( cg, source_ranges, target_ranges );
+
+  cg->start();
 
-  Subnet* targets =
-    dynamic_cast< Subnet* >( kernel().node_manager.get_node( target_id ) );
-  if ( targets == NULL )
+  int source, target, num_parameters = cg->arity();
```

Comment:
Please split into separate defintions. `source` and `target` should probably be `long` or  `index`, while `num_parameters` can be `const`.

Reply:
The split and addition of `const` was done in 43e8d6abf. I'll keep them as `int` as this is what the `ConnectionGenerator` interface defines them as.